dvc/repo/tree.py

```diff
@@ -248,78 +254,148 @@ class RepoTree(BaseTree):  # pylint:disable=abstract-method
     scheme = "local"
     PARAM_CHECKSUM = "md5"
 
-    def __init__(self, repo, **kwargs):
+    def __init__(self, repo, subrepos=False, repo_factory=None, **kwargs):
```

Comment:
`repo_factory` will be needed for `ExternalRepo` as we need to set `cache` and other configurations.

Reply:
Maybe let's start with that first? The current way is confusing, I'm afraid that it will be hard to clean it up once again, as the complexity will be growing more and more.

Reply:
It'd be helpful if you could point out what you find confusing.

Reply:
> Maybe let's start with that first?

ExternalRepo will just use `repo_factory`. This is a requirement for `ExternalRepo`, so there will be no need for cleanups.

Reply:
I have a new alternate design, let's see.